Add DT/DT_INST_CHILD_NUM and DT/DT_INST_CHILD_NUM_STATUS_OKAY for the number of child nodes of a given node

swift-tk opened the pull request:
Background: https://github.com/zephyrproject-rtos/zephyr/issues/38715

The below macro is first proposed but does not work with macros that takes a number as its parameter at pre-processor time.

`#define DT_NUM_CHILD(node_id) (DT_FOREACH_CHILD_SEP_VARGS(node_id, DT_CAT, (+), _EXISTS))`

A better solution is to create a generated macro with the number of child as a number then derive from it.
`#define DT_N(node-path)_CHILD_NUM` 
`#define DT_CHILD_NUM(node_id) DT_CAT(node_id, _CHILD_NUM)` 

Added another generated macro for the number of child which status is okay.
`#define DT_N(node-path)_CHILD_NUM_STATUS_OKAY`
`#define DT_CHILD_NUM_STATUS_OKAY(node_id) DT_CAT(node_id, _CHILD_NUM_STATUS_OKAY)`
This does not conflict with DT_NUM_INST_STATUS_OKAY as it gets the number of status okay of all child nodes instead of with a given compat.

swift-tk committed:
devicetree: Add DT/DT_INST_CHILD_NUM and DT/DT_INST_CHILD_NUM_STATUS_OKAY

Add a generated macro for the number of child nodes of a given node.
Add a generated macro for the number of child nodes of a given node which
children's status are "okay".

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

decsny reviewed (dismissed):
Have been meaning to add this exact macro in this exact way for a while now but keep forgetting

aaronemassey reviewed (dismissed):
LGTM, but get @galak approval.

de-nordic commented:
I wonder whether it would be possible to use MACRO_MAP_CAT_N and https://github.com/zephyrproject-rtos/zephyr/pull/71890 to directly get a number, not equation. That could be directly useful with other macros, as the sum can not be evaluated at preprocessing time.

henrikbrixandersen reviewed (changes requested):
This needs a test case in `tests/lib/devicetree/api`.

swift-tk committed:
tests: devicetree: Add a test test_child_nodes_number

The test uses DT_CHILD_NUM and DT_INST_CHILD_NUM to get the number of
child nodes of a given node from generated macro.
The test uses DT_CHILD_NUM_STATUS_OKAY and DT_INST_CHILD_NUM_STATUS_OKAY
to get the number of child nodes of a given node which children's status
are "okay" from generated macro.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk commented:
> I wonder whether it would be possible to use MACRO_MAP_CAT_N and #71890 to directly get a number, not equation. That could be directly useful with other macros, as the sum can not be evaluated at preprocessing time.

This would work. It is able to work with LISTIFY.

`#define DT_NUM_CHILD(node_id) NUM_VA_ARGS(DT_FOREACH_CHILD_SEP_VARGS(node_id, DT_CAT, (,), _EXISTS))`

But I wonder if we even need it work with macros like LISTIFY. After all, if one needs to use it for iterating, shouldn't they be using DT_FOREACH_CHILD? 
In my context, I'm able to use it in struct initializer macros for DT_DEFINES and it is enough.

decsny commented:
> @decsny @galak Please comment on [#71933 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71933#issuecomment-2078781144). Shall we proceed or close?

Both of them have their limitations I guess, if someone needs something even more robust for some reason then they can submit a later PR to improve it. Whichever you go with, you should put documentation of the macro about it's limitations, ie, will it not work with other macros, or will it not work if a node has more than 64 children. I am not too worried about it since it is a new macro, it's not going to break anything, just make sure to document the limitation so someone isnt confused if they hit the barrier, since macro errors are notoriously confusing

swift-tk commented:
@decsny Thanks.
 
I believe I have found a better solution. By creating a generated macro and derive the children number from it, this will work with macros such as LISTIFY that requires a number at pre-processor time.  

@henrikbrixandersen @galak @decsny Please review again thanks.

pdgendt commented on scripts/dts/gen_defines.py:
```diff
@@ -132,6 +132,9 @@ def main():
                 out_dt_define(f"{node.z_path_id}_CHILD_IDX",
                               node.parent.child_index(node))
 
+                out_comment(f"Node's number of children:")
+                out_dt_define(f"{node.z_path_id}_CHILD_NUM", len(node.children))
```
This should not be part of `if node.parent is not None:`, and could probably move to `write_children`

swift-tk replied:
> This should not be part of `if node.parent is not None:`, and could probably move to `write_children`

May I ask why? All the other things like CHILD_INDEX are under that if statement.

pdgendt replied:
Because `CHILD_IDX` is the current node's index relative to it's parent. The current node's number of children does not depend on having a parent.

swift-tk replied:
> Because `CHILD_IDX` is the current node's index relative to it's parent. The current node's number of children does not depend on having a parent.

The difference is really whether we need a child number for the top node. Adding that outside of the if statement creates a DT_N_CHILD_NUM for the top node. Is that expected?

pdgendt replied:
IMO yes, same as `write_children` is called for the top node.

swift-tk replied:
Done, thanks. Don't approve yet, lets see what others thinks. Plus I will probably need to rebase.

pdgendt reviewed (approved):
I think this is a better approach, thanks!

pdgendt commented on include/zephyr/devicetree.h:
```diff
@@ -3442,6 +3450,17 @@
+ * <tt>DT_CHILD_NUM(DT_DRV_INST(inst))</tt>.
+ *
+ * @param inst Devicetree instance number
+ * @return an array of gpio_dt_spec struct corresponding with mspi_dev's chip enables
```
This `@return` value is wrong.

swift-tk replied:
Fixed.

pdgendt commented on include/zephyr/devicetree.h:
```diff
@@ -551,6 +551,14 @@
+ * @param node_id a node identifier
+ * @return Number of child nodes
+ */
+#define DT_CHILD_NUM(node_id) DT_CAT(node_id, _CHILD_NUM)
```
Perhaps we should align with the `FOREACH_CHILD` macros, and add `DT_CHILD_NUM_STATUS_OKAY` too.

swift-tk replied:
Sure, it has been added.

decsny reviewed (approved):
definitely the correct way to do this now

henrikbrixandersen reviewed (approved):
The PR description needs an update to match the implementation.
